timechop/timechop.py

```diff
@@ -7,169 +7,380 @@
-            test_label_window
-                    ):
-        update_delta = utils.convert_str_to_relativedelta(self.update_window)
-        train_delta = utils.convert_str_to_relativedelta(train_duration)
-        test_label_delta = utils.convert_str_to_relativedelta(test_label_window)
-        matrix_end_times = []
-        matrix_end_time = self.modeling_end_time - test_label_delta
-        
-        print('Initial matrix end time {}'.format(matrix_end_time))
-        if matrix_end_time <= self.modeling_start_time:
-            raise ValueError('No valid training periods in modeling time.')
+            training_prediction_span,
+            test_prediction_span,
+            test_span
+        ):
+        """ Calculate the split times between train and test matrices. All
+        label spans in train matrices will end at this time, and this will be
+        the first as of time in the respective test matrix.
+
+        :param training_prediction_span: how much time is included in training 
+                                         labels
+        :type training_prediction_span: dateutil.relativedelta.relativedelta
+        :param test_prediction_span: how much time is included in test labels
+        :type test_prediction_span: dateutil.relativedelta.relativedelta
+        :param test_span: for how long after the end of a training matrix are
+                          test predictions made
+        :type test_span: str
+
+        :return: all split times for the temporal parameters
+        :rtype: list
+
+        :raises: ValueError if there are no valid split times in the temporal
+                 config
+        """
+        last_test_label_time = self.label_end_time - test_prediction_span
+        # final label must be able to have feature data associated with it
+        if last_test_label_time > self.feature_end_time:
+            last_test_label_time = self.feature_end_time
+            warnings.warn('Final test label date is after end of feature time. Moving labels back to end of feature time.')
+        logging.info('Final label as of date: {}'.format(last_test_label_time))
+
+        # all split times have to allow at least one training label before them
+        earliest_possible_split_time = (training_prediction_span
+            + max(self.feature_start_time, self.label_start_time))
+        logging.info(
+            'Earliest possible train/test split time: {}'.format(
+                earliest_possible_split_time
+            )
+        )
 
-        while matrix_end_time > self.modeling_start_time:
-            matrix_end_times.insert(0, matrix_end_time)
-            matrix_end_time -= update_delta
+        # last split is the first as of time in the final test matrix
+        test_delta = utils.convert_str_to_relativedelta(test_span)
+        test_span_unit_delta = utils.get_one_unit_delta(test_span)
+        last_split_time = (
+            last_test_label_time - test_delta + test_span_unit_delta
```

Comment:
I'm not sure I follow the logic with the test span unit delta here -- won't we get a different result here if a 1 year test span is defined as 1 year, 12 months, 52 weeks, or 365 days?

Reply:
Oh, I figured out what you mean here. I'll need to hand work through some examples to figure out the desired behavior, but the intention here is to resolve the problem that you need to roll back from the sum by some amount to get the last label date because the sum is out of bounds of the matrix.

We could make this hardcoded as 1 day on the assumption that all projects using timechop have one day granularity, but this may not be desirable either.

Reply:
The other option we considered was to make the user specify granularity, but this seemed really hard to communicate to users and for user to figure out for themselves.

Reply:
I think I'm still a little confused, but it might help to walk through an example or something. If it's just about creating a buffer with the granularity of the data, my guess is you might be able to get the same result by changing a `<=` to a `<` in the code somewhere and avoid the problem of being sensitive to units altogether.